File: output/llama3.1-8b/avatar/Java/Python/atcoder_ABC122_D.py
import sys

class Solution:
    def __init__(self):
        self.N = 0
        self.MOD = 1000000007
        self.TOKENS = ["A", "C", "G", "T"]
        self.memo = {}

    def calc(self):
        return self.dfs(0, "TTT")

    def is_ok(self, last4):
        if "AGC" in last4:
            return False
        for i in range(3):
            vals = list(last4)
            vals[i], vals[i+1] = last4[i+1], last4[i]
            s = "".join(vals)
            if "AGC" in s:
                return False
        return True

    def dfs(self, current, last3):
        if (current, last3) in self.memo:
            return self.memo[(current, last3)]
        if current == self.N:
            return 1
        result = 0
        for c in self.TOKENS:
            if self.is_ok(last3 + c):
                result += self.dfs(current+1, last3[1:] + c)
                result %= self.MOD
        self.memo[(current, last3)] = result
        return result

if __name__ == "__main__":
    ins = Solution()
    ins.N = int(input())
    print(ins.calc())

File: output/llama3.1-8b/avatar/Python/Java/atcoder_ABC122_D.java
import java.util.HashMap;
import java.util.Map;

public class atcoder_ABC122_D {
    public static boolean ok(String last4) {
        for (int i = 0; i < 4; i++) {
            char[] t = last4.toCharArray();
            if (i >= 1) {
                char temp = t[i];
                t[i] = t[i - 1];
                t[i - 1] = temp;
            }
            String s = new String(t);
            if (s.contains("AGC")) {
                return false;
            }
        }
        return true;
    }

    public static int dfs(int cur, String last3) {
        Map<String, Integer> memo = getMemo(cur);
        if (memo.containsKey(last3)) {
            return memo.get(last3);
        }
        if (cur == n) {
            return 1;
        }
        int ret = 0;
        for (char c : "ACGT".toCharArray()) {
            String newLast3 = last3 + c;
            if (ok(newLast3)) {
                ret = (ret + dfs(cur + 1, newLast3.substring(1))) % mod;
            }
        }
        memo.put(last3, ret);
        return ret;
    }

    public static Map<String, Integer> getMemo(int n) {
        Map<String, Integer>[] memo = new HashMap[n + 1];
        for (int i = 0; i <= n; i++) {
            memo[i] = new HashMap<>();
        }
        return memo[0];
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(System.console().readLine());
        int mod = (int) Math.pow(10, 9) + 7;
        Map<String, Integer>[] memo = getMemo(n);
        System.out.println(dfs(0, "TTT"));
    }
}